pkg/timer/timer.go

```diff
@@ -0,0 +1,93 @@
+package timer
+
+import (
+	"context"
+	"github.com/filecoin-project/mir/pkg/events"
+	t "github.com/filecoin-project/mir/pkg/types"
+	"time"
+)
+
+// The Timer module abstracts the passage of real time.
+// It is used for implementing timeouts and periodic repetition.
+type Timer struct {
+	retIndex t.TimerRetIndex
+}
+
+// Delay writes events to notifyChan after delay.
+// If context is canceled, no events might be written.
+func (t *Timer) Delay(
+	ctx context.Context,
+	events *events.EventList,
+	delay t.TimeDuration,
+	notifyChan chan<- *events.EventList,
+) {
+
+	// This manual implementation has the following advantage over simply using time.AfterFunc:
+	// When the context is canceled, the timer stops immediately
+	// and does not still fire after a (potentially long) delay.
+	go func() {
+		select {
+		case <-time.After(time.Duration(delay)):
+			// If the delay elapses before the context is canceled,
+			// try to write events to notifyChan until the context is canceled.
+			select {
+			case notifyChan <- events:
+			case <-ctx.Done():
+			}
+		case <-ctx.Done():
+			// If the context is canceled before the delay elapses, do nothing.
+		}
+	}()
+}
+
+// Repeat periodically writes events to notifyChan with the given period.
+// The repeated producing of events is associated with the retention index retIndex.
+// The first write of events to notifyChan happens immediately after invocation of Repeat.
+// Repeat stops writing events to notifyChan when
+// 1) ctx is canceled or
+// 2) GarbageCollect is called with an argument greater than the associated retIndex.
+func (t *Timer) Repeat(
+	ctx context.Context,
+	events *events.EventList,
+	period t.TimeDuration,
+	retIndex t.TimerRetIndex,
+	notifyChan chan<- *events.EventList,
+) {
+	go func() {
+
+		// Repeat as long as this repetition task has not been garbage-collected.
+		for retIndex >= t.retIndex {
+
+			// Try to write events to notifyChan until the context is canceled.
+			select {
+			case notifyChan <- events:
+			case <-ctx.Done():
+			}
+
+			// Wait for the defined period of time (or until the context is canceled)
+			select {
+			case <-time.After(time.Duration(period)):
```

Comment:
I think it is better to use a channel returned from [`time.Tick`](https://pkg.go.dev/time#Tick), because otherwise the interval may not be approximately equal to `period`.

Reply:
This approach will send events with interval strictly grater than `duration` because `time.After` cannot close the returned channel before the passed duration elapses (in practice it will always take a little longer) and the other select statement in the loop may also take some time. 

> Tick ... [leaks memory](https://pkg.go.dev/time#Tick)

You're right, let's use `time.Ticker` instead.

Reply:
If we use `time.Ticker` then we'll have events sent with interval `period` on average.